cypress/integration/unit_tests/checkbox_test.js

```diff
@@ -92,35 +101,238 @@ describe('Unit test for toggleLayerOn', () => {
   });
 
   it('check hidden layer, then uncheck before list evaluation', () => {
-    expect(layerArray[2].displayed).to.equals(false);
+    expect(layerArray[2].displayed).to.be.false;
     expect(layerArray[2].data).to.be.undefined;
 
     const emptyList = [];
     let callback = null;
     stubForEmptyList((callb) => callback = callb);
-    toggleLayerOn(mockFirebaseLayers[2]);
+    const promise = toggleLayerOn(mockFirebaseLayers[2]);
+    expect(promise).to.not.be.null;
+    expect(layerArray[2].displayed).to.be.true;
+    expect(layerArray[2].data).to.be.undefined;
     toggleLayerOff(2);
     callback(emptyList);
-
-    // Evaluate after the promise finishes by using an instant wait.
-    cy.wait(0).then(() => {
-      expect(layerArray[2].displayed).to.equals(false);
+    cy.wrap(promise).then(() => {
+      expect(layerArray[2].displayed).to.be.false;
       expect(layerArray[2].data).to.not.be.null;
       const layerProps = deckGlArray[2].props;
       expect(layerProps).to.have.property('id', 'asset2');
       expect(layerProps).to.have.property('visible', false);
       expect(layerProps).to.have.property('data', emptyList);
     });
   });
+
+  it('checks hidden layer, unchecks, checks before list evaluation', () => {
+    expect(layerArray[2].displayed).to.be.false;
+    expect(layerArray[2].data).to.be.undefined;
+
+    const emptyList = [];
+    let callback = null;
+    stubForEmptyList((callb) => callback = callb);
+    const promise = toggleLayerOn(mockFirebaseLayers[2]);
+    expect(promise).to.not.be.null;
+    expect(layerArray[2].displayed).to.be.true;
+    expect(layerArray[2].data).to.be.undefined;
+    toggleLayerOff(2);
+    const secondPromise = toggleLayerOn(mockFirebaseLayers[2]);
+    expect(layerArray[2].displayed).to.be.true;
+    expect(layerArray[2].data).to.be.undefined;
+    expect(secondPromise).equals(promise);
+    callback(emptyList);
+    cy.wrap(promise).then(() => {
+      expect(layerArray[2].displayed).to.be.true;
+      expect(layerArray[2].data).to.not.be.null;
+      const layerProps = deckGlArray[2].props;
+      expect(layerProps).to.have.property('id', 'asset2');
+      expect(layerProps).to.have.property('visible', true);
+      expect(layerProps).to.have.property('data', emptyList);
+    });
+  });
+
+  // For the next three tests, we do the following setup:
+  // 1. Use a real map, since we want to see that it has an entry in its
+  // overlayMapTypes.
+  // 2. Sub in trivial image, and control the #getMap method of that image so
+  // that we can delay the callback until we're ready.
+  // 3. Stub the loading elements, so we can check when loading starts/ends.
+  it('caches computed image overlay and starts loading on EE request', () => {
+    const map = createGoogleMap();
```

Comment:
functionalize?

Reply:
Not that neat, since we need to return all four things, seems messy. Also, this is refactored in the follow-up.